### PR TITLE
Destroy Context* throughout MonClient. Also do other things.

### PR DESCRIPTION
**Pull request by adamemerson**
Also included is making everything more RAIIful, using an intrusive map in one place, and other general cleanups.

### CODE REVIEW

File: src/mon/MonClient.cc [hunk 1]
```diff
@@ -397,13 +369,13 @@ int MonClient::init()
...
-  rotating_secrets = new RotatingKeyRing(cct, cct->get_module_type(), keyring);
+  rotating_secrets.reset(new RotatingKeyRing(
+			   cct, cct->get_module_type(), keyring.get()));
 
   initialized = true;
 
-  timer.init();
   finisher.start();
-  schedule_tick();
+  timer.add_event(tick_time(), &MonClient::timer, this);
```
Comment (cbodley) [inline src/mon/MonClient.cc]: SafeTimer doesn't spawn a thread until `init()`, so this timer should probably start suspended as well.

Comment (adamemerson) [inline src/mon/MonClient.cc]: I've been trying to move away from the pattern of having 'init' and 'shutdown' methods separate from constructors and destructors, so if there isn't a compelling reason to do it that way, I'd rather not.

If anything I'd rather throw a patch in that gets rid of MonClient's init and shutdown methods. (But I was being conservative about changes to the interface, apart from removing Context*.)

Comment (cbodley) [inline src/mon/MonClient.cc]: fair enough. just pointing out that the timer's initialization is not consistent with the rest of MonClient

File: src/common/waiter.h [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	unique_lock l(exec_base());
+	ret = std::forward_as_tuple(_ret...);
+      }
+    };
+  }
+
+  using waiter_detail::waiter;
+}
+
+#endif // CEPH_COMMON_WAITER_H
```
Comment (cbodley) [inline src/common/waiter.h]: very elegant. so much better than having to subclass C_SafeCond

File: src/mon/MonClient.cc (hunk 1 above)
Comment (chardan) [inline src/mon/MonClient.cc]: +1 ctors/dtors when separate declaration and use isn't really needed. My $0.25, anyway! :-)

File: src/mon/MonClient.cc [hunk 3]
```diff
@@ -220,35 +184,40 @@ int MonClient::ping_monitor(const string &mon_id, string *result_reply)
...
 
-  MonClientPinger *pinger = new MonClientPinger(cct, result_reply);
+  std::unique_ptr<MonClientPinger> pinger(
+    new MonClientPinger(cct, result_reply));
+
 
-  Messenger *smsgr = Messenger::create_client_messenger(cct, "temp_ping_client");
-  smsgr->add_dispatcher_head(pinger);
+  auto dm = [this](Messenger* m) {
+    m->shutdown();
```
Comment (chardan) [inline src/mon/MonClient.cc]: Is m guaranteed to be valid at this point? It's not the case in the code I see here, but is there any chance of the unique_ptr<> ever being default constructed and hitting this Deleter?

**Comment by chardan** [comment]
I'm new to Ceph so it's hard to comment on the behavior component, but I like this patch set. Nice work!

File: src/mon/MonClient.cc (hunk 3 above)
Comment (adamemerson) [inline src/mon/MonClient.cc]: Yes. Also in the case of a capturing lambda as the deleter, the unique_ptr simply can't be default constructed. (That's why I have to initialize it specifically with nullptr and dm.)

**Comment by liewegas** [comment]
rados suite crashed and burned: http://pulpito.ceph.com/sage-2016-02-25_11:08:41-rados-wip-sage-testing---basic-smithi/

File: src/common/Finisher.cc [hunk 4]
```diff
@@ -12,96 +14,93 @@
...
-	  // When an item is NULL in the finisher_queue, it means
-	  // we should instead process an item from finisher_queue_rval,
-	  // which has a parameter for complete() other than zero.
-	  // This preserves the order while saving some storage.
-	  assert(!ls_rval.empty());
-	  Context *c = ls_rval.front().first;
-	  c->complete(ls_rval.front().second);
-	  ls_rval.pop_front();
-	}
+      for (auto& p : ls) {
```
Comment (yuyuyu101) [inline src/common/Finisher.cc]: why not for (auto &&p : ls) {
  p();

so we can reduce a move?

**Comment by athanatos** [comment]
@adamemerson Did the above rebase fix the crash?

**Comment by adamemerson** [comment]
@athanatos Don't know yet. I'm waiting for gitbuilder to gitbuild it and I'll try re-running the test on Sepia.

**Comment by athanatos** [comment]
@adamemerson Let me know when you want me to pull it into my integration branch.

**Comment by tchaikov** [comment]
ping @adamemerson

**Comment by liewegas** [comment]
@adamemerson @tchaikov i suspect this needs a massive rebase.. and it probably conflicts terribly with the monclient refactor to have parallel sessions?

**Comment by dachary** [comment]
feel free to re-open if you think this still is relevant

**Pull request closed without merging**